### PR TITLE
Add pip compile

### PR DESCRIPTION
**Pull request by neob91-close**
I just wanted to introduce `pip-compile`, but had to switch to `ruff` as well, because we had some conflicting linting packages, and after fixing that, flake8 started reporting a whole lot of different errors.

We were going to introduce `ruff` anyway, I just wish we didn't have to do it all in one PR.

Definitely review commits separately.

### CODE REVIEW

File: .github/workflows/build-and-release.yaml [hunk 1]
```diff
@@ -40,6 +40,34 @@ jobs:
       - name: Run flake8
         run: flake8
 
+  check-compiled-requirements:
```
Comment (neob91-close) [inline .github/workflows/build-and-release.yaml:40]: Mostly copied from `closeio`

**Comment by neob91-close** [comment]
Now I'm thinking maybe I should just swap out the current `*.txt` requirements files for the autogenerated ones as a separate PR.

**Review by squeaky-pl — APPROVED** [review]
The general structure looking good. Not reviewing machine generated changes.

**Comment by tsx** [comment]
>  I just wish we didn't have to do it all in one PR.

What's stopping you from shipping ruff changes separately first?

Doing both ruff and dependency resolver change seems like an insane amount of risk to me.

**Comment by neob91-close** [comment]
Ruff PR: https://github.com/closeio/sync-engine/pull/985

**Comment by neob91-close** [comment]
Removed the ruff-related changes, and regenerated the requirements files.

**Pull request merged**